File: Laboratory work #1/Python/Gauss.py
import numpy as np
from sys import argv

class GaussMethod :

    def __init__ (self) :
        self.systemSize = 0
        self.coeff = []
        self.freeTerm = []
        self.solutions = []
        self.filename = "matrix.txt"
        self.isSolutionFounded = False

    def initializationFromUserInput(self):
        self.systemSize = int(input("Р’РІРµРґРёС‚Рµ РєРѕР»РёС‡РµСЃС‚РІРѕ СѓСЂР°РІРЅРµРЅРёР№: "))

        print("Р’РІРµРґРёС‚Рµ РєРѕСЌС„С„РёС†РёРµРЅС‚С‹: ")
        for i in range(self.systemSize) :
            print(f"{i+1} СѓСЂР°РІРЅРµРЅРёРµ: ")
            self.coeff.append([])

            for j in range(self.systemSize) :
                self.coeff[i].append(float(input(f"a[{i}][{j}]= ")))

            print("Р’РІРµРґРёС‚Рµ СЃРІРѕР±РѕРґРЅС‹Р№ С‡Р»РµРЅ: ")
            self.freeTerm.append(float(input(f"y[{i}]= ")))

        # print("Р’РІРµРґРёС‚Рµ СЃРІРѕР±РѕРґРЅС‹Р№ С‡Р»РµРЅ: ")
        # for i in range(self.systemSize) :
        #     self.freeTerm.append(int(input(f"y[{i}]= ")))

        self.solutions = [None for i in range(0,self.systemSize)]

    def initializationFromFile(self) :
        with open('matrix.txt') as file :
            self.coeff = [list(map(float, row.split())) for row in file.readlines()]
            self.systemSize = len(self.coeff[0])-1

        for i in range(0,self.systemSize) :
            self.freeTerm.append(self.coeff[i][-1])
            self.coeff[i].pop()

        self.solutions = [None for i in range(0,self.systemSize)]

    def showSystem(self):
        print("Р’РІРµРґРµРЅРЅР°СЏ СЃРёСЃС‚РµРјР°: ")
        for row in range(len(self.freeTerm)):
            for col in range(len(self.coeff[row])):
                print(f"{self.coeff[row][col]}*x{col}", end='')
                if col < self.systemSize - 1 :
                    print(" + ", end='')
            print(f" = {self.freeTerm[row]}")

    def applyGaussMethod (self):
        x = [0 for i in range(0,self.systemSize)]

        tmpNpArr = np.array(self.coeff)
        rankOfUsualMatrix = np.linalg.matrix_rank(tmpNpArr)
        # determinator = np.linalg.det(self.coeff)

        for i in range(0,self.systemSize) :
            self.coeff[i].append(self.freeTerm[i])

        tmpNpArr = np.array(self.coeff)
        rankOfExtendedMatrix = np.linalg.matrix_rank(tmpNpArr)

        if rankOfUsualMatrix < rankOfExtendedMatrix :
            print("РќРµС‚ СЂРµС€РµРЅРёСЏ!")
            return
        elif rankOfUsualMatrix == rankOfExtendedMatrix and rankOfUsualMatrix < self.systemSize :
            print("РЎРёСЃС‚РµРјР° РёРјРµРµС‚ Р±РµСЃРєРѕРЅРµС‡РЅРѕРµ РјРЅРѕР¶РµСЃС‚РІРѕ СЂРµС€РµРЅРёР№")
            return
        elif rankOfUsualMatrix == rankOfExtendedMatrix and rankOfUsualMatrix == self.systemSize :
            print("РЎРёСЃС‚РµРјР° РёРјРµРµС‚ РµРґРёРЅСЃС‚РІРµРЅРЅРѕРµ СЂРµС€РµРЅРёРµ!")

        for i in range(self.systemSize-1,0,-1) :
            if self.coeff[i-1][0] < self.coeff[i][0] :
                for j in range(0,self.systemSize+1) :
                    self.coeff[i][j], self.coeff[i-1][j] = self.coeff[i-1][j], self.coeff[i][j]

        for k in range(0,self.systemSize-1) :
            for i in range(k,self.systemSize-1) :
                tmp = (self.coeff[i+1][k] / self.coeff[k][k])

                for j in range(0,self.systemSize+1) :
                    self.coeff[i+1][j] -= tmp * self.coeff[k][j]

        for i in range(self.systemSize-1,-1,-1) :
            tmp = 0.0
            for j in range(i,self.systemSize) :
                tmp = tmp + self.coeff[i][j]*x[j]
            x[i] = (self.coeff[i][self.systemSize]-tmp) / self.coeff[i][i]

        self.isSolutionFounded = True

        for i in range(0,self.systemSize) :
            self.solutions[i] = x[i];
        # self.solutions = x

    def showSolutions(self) :
        if self.isSolutionFounded :
            print("Р РµС€РµРЅРёСЏ РІРІРµРґРµРЅРЅРѕР№ СЃРёСЃС‚РµРјС‹: ");
            for i in range(self.systemSize) :
                print(f"x[{i}]= {self.solutions[i]}")

if __name__ == "__main__" :
    scriptName, mode = argv

    gaussMethod = GaussMethod()
    if mode == 'user' :
        gaussMethod.initializationFromUserInput()
    elif mode == 'file' :
        gaussMethod.initializationFromFile()
    gaussMethod.showSystem()
    gaussMethod.applyGaussMethod()
    gaussMethod.showSolutions()

File: Laboratory work #1/C++/Gauss.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <math.h>
#include <utility>
#include <Eigen/Dense>

using namespace Eigen;
using namespace std;

class GaussMethod{
private:
    int systemSize;
    double** coeff;
    double** tmpMatrix;
    double* freeTerm;
    double* solutions;
    string filename;
    bool isSolutionFounded;
public:
    GaussMethod();
    ~GaussMethod();
    long int getRank(double**, int, int);
    void swap(double**, int, int, int);
    void initializeFromFile();
    void initializeFromUserInput();
    void showSystem();
    void applyGausMethod();
    void showSolutions();
    void deleteTmp();
    double** getCopy(bool);
};

GaussMethod::GaussMethod()
{
    filename = "matrix.txt";
    isSolutionFounded = false;
}

GaussMethod::~GaussMethod()
{
    delete [] solutions;
    for (int i = 0; i < systemSize; i++)
        delete [] coeff[i];
    delete [] coeff;
    delete [] freeTerm;
}

long int GaussMethod::getRank(double** a, int nSize, int mSize)
{
    MatrixXd matrix(nSize, mSize);
    for(int i = 0; i < nSize; i++)
    {
        for(int j = 0; j < mSize; j++)
        {
            matrix(i,j) = a[i][j];
        }
    }

    FullPivLU<MatrixXd> lu_decomp(matrix);
    long int rank = lu_decomp.rank();

    return rank;
}

void GaussMethod::swap(double** mat, int row1, int row2, int col)
{
    for (int i = 0; i < col; i++)
        {
            double temp = mat[row1][i];
            mat[row1][i] = mat[row2][i];
            mat[row2][i] = temp;
        }
}

void GaussMethod::initializeFromFile()
{
    ifstream fout(filename);

    fout >> systemSize;

    coeff = new double*[systemSize];
    freeTerm = new double[systemSize];

    for (int i = 0; i < systemSize; i++)
    {
        coeff[i] = new double[systemSize+1];
        for (int j = 0; j < systemSize; j++)
        {
            fout >> coeff[i][j];
        }

        fout >> freeTerm[i];
    }

    fout.close();
}

void GaussMethod::initializeFromUserInput()
{
    cout << "Р’РІРµРґРёС‚Рµ РєРѕР»РёС‡РµСЃС‚РІРѕ СѓСЂР°РІРЅРµРЅРёР№: ";
    cin >> systemSize;
    coeff = new double*[systemSize];
    freeTerm = new double[systemSize];

    cout << "Р’РІРµРґРёС‚Рµ РєРѕСЌС„С„РёС†РёРµРЅС‚С‹: \n";
    for (int i = 0; i < systemSize; i++)
    {
        cout << i+1 << " СѓСЂР°РІРЅРµРЅРёРµ: \n";
        coeff[i] = new double[systemSize];
        for (int j = 0; j < systemSize; j++)
        {
            cout << "a[" << i << "][" << j << "]= ";
            cin >> coeff[i][j];
        }

        cout << "Р’РІРµРґРёС‚Рµ СЃРІРѕР±РѕРґРЅС‹Р№ С‡Р»РµРЅ: \n";
        cout << "y[" << i << "]= ";
        cin >> freeTerm[i];
    }

}

void GaussMethod::showSystem()
{
    cout << "Р’РІРµРґРµРЅРЅР°СЏ СЃРёСЃС‚РµРјР°: \n";
    for (int i = 0; i < systemSize; i++)
    {
        for (int j = 0; j < systemSize; j++)
        {
            if (coeff[i][j] < 0)
                cout << "(" << coeff[i][j] << ")*x" << j;
            else
                cout << coeff[i][j] << "*x" << j;
            if (j < systemSize - 1)
                cout << " + ";
        }
        cout << " = " << freeTerm[i] << endl;
    }
    return;
}

void GaussMethod::applyGausMethod()
{
    solutions = new double[systemSize];

    long int rankOfUsualMatrix = getRank(getCopy(0),systemSize, systemSize);
    deleteTmp();
//    cout << rankOfUsualMatrix << endl;

    double x[systemSize];

    for (int i=0; i<systemSize; i++)
        coeff[i][systemSize] = freeTerm[i];

    long int rankOfExtendedMatrix = getRank(getCopy(1),systemSize,systemSize+1);
    deleteTmp();
//    cout << rankOfExtendedMatrix << endl;

    if (rankOfUsualMatrix < rankOfExtendedMatrix)
    {
        cout << "РќРµС‚ СЂРµС€РµРЅРёСЏ!" << endl;
        return;
    }
    else if (rankOfUsualMatrix == rankOfExtendedMatrix && rankOfUsualMatrix < systemSize)
    {
        cout << "РЎРёСЃС‚РµРјР° РёРјРµРµС‚ Р±РµСЃРєРѕРЅРµС‡РЅРѕРµ РјРЅРѕР¶РµСЃС‚РІРѕ СЂРµС€РµРЅРёР№" << endl;
        return;
    }
    else if (rankOfUsualMatrix == rankOfExtendedMatrix && rankOfUsualMatrix == systemSize)
        cout << "РЎРёСЃС‚РµРјР° РёРјРµРµС‚ РµРґРёРЅСЃС‚РІРµРЅРЅРѕРµ СЂРµС€РµРЅРёРµ!" << endl;

    for(int i=systemSize-1; i>0; i--)
    {
        if(coeff[i-1][0]<coeff[i][0])
        {
            for(int j=0; j<=systemSize; j++)
            {
                double c = coeff[i][j];
                coeff[i][j] = coeff[i-1][j];
                coeff[i-1][j] = c;
            }
        }
    }

//    for (int i = 0; i< systemSize; i++)
//        for (int j = 0; j< systemSize; j++)
//            cout << coeff[i][j];

    for(int k=0; k<systemSize-1; k++)
        for(int i=k; i<systemSize-1; i++)
        {
            double c = (coeff[i+1][k]/coeff[k][k]) ;

            for(int j=0; j<=systemSize; j++)
                coeff[i+1][j]-=c*coeff[k][j];
        }

//    cout << endl;
//    for (int i = 0; i< systemSize; i++)
//        for (int j = 0; j< systemSize; j++)
//            cout << coeff[i][j];

    for(int i=systemSize-1; i>=0; i--)
    {
        double c = 0;
        for(int j=i; j<=systemSize-1; j++)
            c = c+coeff[i][j]*x[j];

        x[i]=(coeff[i][systemSize]-c)/coeff[i][i];
//        cout << x[i] << endl;
    }

    isSolutionFounded = true;

    for (int i=0; i<systemSize; i++)
        solutions[i] = x[i];

//    delete [] x;
}

void GaussMethod::showSolutions()
{
    if (isSolutionFounded)
    {
        cout << "Р РµС€РµРЅРёРµ РІРІРµРґРµРЅРЅРѕР№ СЃРёСЃС‚РµРјС‹: \n";
        for (int i = 0; i < systemSize; i++)
            cout << "x[" << i << "]=" << solutions[i] << endl;
    }
}

void GaussMethod::deleteTmp()
{
    for (int i = 0; i < systemSize; i++)
        delete [] tmpMatrix[i];
    delete [] tmpMatrix;
}

double **GaussMethod::getCopy(bool mode)
{
    tmpMatrix = new double*[systemSize];
    int nSize = 0;
    int mSize = systemSize;
    mode ? nSize = systemSize + 1 : nSize = systemSize;

    for (int i = 0; i < mSize; i++)
    {
        tmpMatrix[i] = new double[nSize];
        for (int j = 0; j < nSize; j++)
            tmpMatrix[i][j] = coeff[i][j];
    }

    return tmpMatrix;
}

int main(int arg, char** argv)
{
    GaussMethod gaussMethod;

    if(strcmp(argv[1],"user") == 0)
        gaussMethod.initializeFromUserInput();
    else if (strcmp(argv[1],"file") == 0)
        gaussMethod.initializeFromFile();

    gaussMethod.showSystem();
    gaussMethod.applyGausMethod();
    gaussMethod.showSolutions();

    return 0;
}
